### ISSUE TITLE
Potential issue with YAML variables containing dashes

### ISSUE THREAD

**Issue by erikmd** [coq-community/templates#14]
Regarding the possible implementations of Mustache, I've successively tested pystache and mo.

On the one hand, pystache seems to be broken:

```
$ sudo pip3 install pystache

$ pystache meta.yml ../coq-community-templates/.travis.yml.mustache
Traceback (most recent call last):
  File "/usr/local/bin/pystache", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.5/dist-packages/pystache/commands/render.py", line 86, in main
    context = json.load(open(context))
  File "/usr/lib/python3.5/json/__init__.py", line 268, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib/python3.5/json/__init__.py", line 319, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.5/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
$ 
```

but on the other hand, `mo` (the Bash implementation) seems not to recognize YAML variables containing dashes:

```
$ cd .../coqoban

$ grep mo ../coq-community-templates/generate.sh 
    mo meta.yml "$f" > "$(basename "$f" .mustache)"

$ ../coq-community-templates/generate.sh 
Generating coq.opam...
/usr/local/bin/mo: line 988: opam-file-maintainer: bad substitution
Generating default.nix...
Generating extracted.opam...
Generating index.md...
Generating README.md...
Generating .travis.yml...
```

hence this issue: should the `-`'s:

https://github.com/coq-community/templates/blob/84a16ca1931c70eb3541492257335fac0ad51e22/coq.opam.mustache#L2

be replaced with `_`'s like in this other line?

https://github.com/coq-community/templates/blob/84a16ca1931c70eb3541492257335fac0ad51e22/coq.opam.mustache#L18

Finally out of curiosity @palmskog and @Zimmi48, which implementation of mustache are you using?

**Comment by palmskog** [comment]
I'm using the Ruby mustache implementation. Not allowing dashes is quite a big limitation in my book, can you confirm that some other implementation besides the bash one disallows this?

I'm not completely averse to switching to underscores though, since it only seems to be a few fields that use this.

**Comment by erikmd** [comment]
Thanks @palmskog, I confirm it was very fine with the Ruby implementation

(for the record, it amounted to installing this APT package: `apt-get install ruby-mustache`)

> Not allowing dashes is quite a big limitation in my book, can you confirm that some other implementation besides the bash one disallows this?

I've only done a few experiments with Bash, Python and Ruby so I can't really confirm this (but BTW it appears Bash and Python do not take a YAML as input: Bash takes environment variables (hence the `-` limitation in the template syntax it accepts I guess) and Python takes a JSON format).

So maybe this is not really an issue, but indeed:

* you might want to switch to underscores for uniformity(?)
* and in any case, it may be useful to recommend the mustache implementation somewhere in the templates repo(?) [all the more as [it was the initial implementation](https://en.wikipedia.org/wiki/Mustache_(template_system)#History_and_principles)]

**Comment by palmskog** [comment]
@erikmd I'm open to switching to "underscores only" in all templates, feel free to do a PR with this. You can also add something to README about the Ruby mustache implementation.

**Comment by erikmd** [comment]
@palmskog 
> You can also add something to the README about the Ruby mustache implementation.

done: https://github.com/coq-community/templates/pull/17

>  I'm open to switching to "underscores only" in all templates, feel free to do a PR with this.

On second thought, I am unsure it is very practical to do such a refactoring, at least now, as the templates repo only contain the `.mustache` files but *not* an authoritative sample `meta.yml` file. 

I don't know either if adding a sample `meta.yml` with underscores in the templates repo would be a solution, as in any case changing the naming convention of the mustache variables would require to change all `meta.yml` files in the coq-community projects (not only the three mentioned in the templates README) to avoid breaking future updates by these projects' maintainers.

So to sum up: the issue I raised would be only blocking when using these `.mustache` files with the environment-variables based Bash mustache implementation, but as the overall recommendation is to use a `.yml` metadata files, maybe we can just keep all this as is − and possibly decide to use underscore for future new variables.

(so, feel free to close this issue if need be)

**Comment by palmskog** [comment]
Not having an authoritative `meta.yml` was a conscious choice, since I would likely have to update it quite often. See coq-community/manifesto#48 for some ideas on how the metadata format could be specified.

**Comment by erikmd** [comment]
OK @palmskog, thanks for the link.

BTW when you say

> would likely have to update it quite often.

I agree with you, and I guess furthermore it's an argument for not changing the metadata items names for `-`/`_` in particular, given that this would imply even more updates for existing `meta.yml` :)

**Comment by Zimmi48** [comment]
FTR I use the Go implementation, which also supports YAML input files.
